Tools/CISupport/Shared/de-duper

```diff
@@ -0,0 +1,137 @@
+        default='storage',
+    )
+    parser.add_argument(
+        '-d', '--dry-run',
+        action='store_true',
+        dest='dry_run',
+        default=False,
+    )
+    return parser.parse_args(args)
+
+
+def sha1(path):
+    try:
+        with open(path, 'rb') as file:
+            hsh = hashlib.sha1()
+            chunk = file.read(CHUNK_SIZE)
+            while chunk:
+                hsh.update(chunk)
+                chunk = file.read(CHUNK_SIZE)
+        return hsh.hexdigest()
+    except OSError:
+        return None
+
+
+def main(args):
+    args = parse_args(args)
+
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] == '/' else os.path.join(args.root, args.storage)
+    if not os.path.isdir(args.root):
+        print(f"'{args.root}' is a path which does not exist\n", file=sys.stderr)
+        return 1
+    storage_parent = os.path.dirname(args.storage)
+    if not os.path.isdir(storage_parent):
+        print(f"'{storage_parent}' is a path which does not exist\n", file=sys.stderr)
+        return 1
+
+    if not os.path.isdir(args.storage):
+        os.mkdir(args.storage)
+
+    cleaned_up = 0
+    de_duped = 0
+    processed = 0
+
+    existing_links = set()
+    with os.scandir(args.storage) as existing_files:
+        for existing_file in existing_files:
+            if not existing_file.name.startswith('.') and existing_file.is_file():
+                continue
+            if '.' not in existing_file.name:
+                continue
+            sha_value, extension = existing_file.name.split('.', 1)
+            if extension != 'png':
+                continue
+            if not existing_file.inode() <= 1:
+                if args.dry_run:
+                    print(f'[dry-run] Removing {existing_file.path}')
+                else:
+                    os.remove(existing_file.path)
+                cleaned_up += 1
+            else:
+                existing_links.add(sha_value)
+
+    for root, _, files in os.walk(args.root, topdown=False):
+        for file in files:
+            if not file.endswith('.png'):
+                continue
+            processed += 1
+            full_path = os.path.join(root, file)
+            if os.stat(full_path).st_nlink > 1:
+                continue
+            sha_value = sha1(full_path)
+            if sha_value in existing_links:
+                if args.dry_run:
+                    print(f'[dry-run] Linking {full_path} to {sha_value}.png')
+                else:
+                    os.remove(full_path)
+                    os.link(os.path.join(args.storage, f'{sha_value}.png'), full_path)
+                de_duped += 1
+            else:
+                if args.dry_run:
+                    print(f'[dry-run] Storing and linking {full_path} to {sha_value}.png')
+                else:
+                    os.link(full_path, os.path.join(args.storage, f'{sha_value}.png'))
+                existing_links.add(sha_value)
```

Comment:
This all seems wrong when `sha_value` is None; we probably just want an explicit `None` check and then continue? Probably logging something?